cmd/sign-off-checker/main.go

```diff
@@ -17,129 +17,198 @@ limitations under the License.
+	invalid := func(msg string, args ...interface{}) {
+		valid = false
+		fmt.Fprintf(os.Stderr, msg+"\n", args...)
 	}
 
-	ts := oauth2.StaticTokenSource(
-		&oauth2.Token{AccessToken: token},
-	)
-	tc := oauth2.NewClient(oauth2.NoContext, ts)
-	client = github.NewClient(tc)
+	if !viper.IsSet("sharedSecret") {
+		invalid("$SHARED_SECRET is not set")
+	}
 
-	log.Print("Starting serving /webhook on :8080")
-	http.Handle("/webhook", loggingMiddleware(http.HandlerFunc(HandleHook)))
-	err := http.ListenAndServe(":8080", nil)
-	log.Fatal(err)
-}
+	if !viper.IsSet("githubToken") {
+		invalid("$GITHUB_TOKEN is not set")
+	}
 
-func HandleHook(w http.ResponseWriter, r *http.Request) {
-	payload, err := github.ValidatePayload(r, secret)
+	_, _, err := net.SplitHostPort(viper.GetString("listenAddress"))
 	if err != nil {
-		http.Error(w,
-			fmt.Sprintf("Could not validate signature: %v", err),
-			http.StatusBadRequest)
-		return
+		invalid("listen address is invalid (%v)", err)
 	}
 
-	hooktype := github.WebHookType(r)
-	event, err := github.ParseWebHook(hooktype, payload)
-	if err != nil {
-		http.Error(w,
-			fmt.Sprintf("Error parsing payload: %v", err),
-			http.StatusBadRequest)
-		return
+	if viper.GetString("publicWebhookURL") != "" {
+		url, err := url.ParseRequestURI(viper.GetString("publicWebhookURL"))
+		if err != nil {
+			invalid("public webhook URL is invalid (%v)", err)
+		} else if url.Scheme != "https" {
+			invalid("public webhook URL must be \"https://[...]\"")
+		}
+	} else if len(viper.GetStringSlice("autoregisterOrganizations")) > 0 {
+		invalid("--public-webhook-url/$PUBLIC_WEBHOOK_URL must be set to use automatic registration")
 	}
-	switch event := event.(type) {
-	case *github.PullRequestEvent:
-		HandlePullRequest(event)
-	default:
-		log.Printf("Unhandled hook type: %v", hooktype)
+
+	if !valid {
+		fmt.Fprintf(os.Stderr, "\n")
+		return fmt.Errorf("invalid parameters")
 	}
+	return nil
 }
 
-func HandlePullRequest(event *github.PullRequestEvent) {
-	owner := event.Repo.Owner.Login
-	repo := event.Repo.Name
-	number := event.Number
+// run the webhook server and autoregistration daemon
+func run() error {
+	gh := github.NewClient(
+		oauth2.NewClient(oauth2.NoContext,
+			oauth2.StaticTokenSource(
+				&oauth2.Token{AccessToken: viper.GetString("githubToken")},
+			),
+		),
+	)
 
-	opt := &github.ListOptions{PerPage: 10}
-	allCommits := []*github.RepositoryCommit{}
-	for {
-		commits, resp, err := client.PullRequests.ListCommits(context.TODO(), *owner, *repo, *number, opt)
-		if err != nil {
-			log.Printf("Error getting commits for PR: %v", err)
-			return
-		}
-		allCommits = append(allCommits, commits...)
-		if resp.NextPage == 0 {
-			break
-		}
-		opt.Page = resp.NextPage
-	}
+	go autoregister(gh)
 
-	signMissing := false
-	for _, commit := range allCommits {
-		if !testRE.MatchString(*commit.Commit.Message) {
-			signMissing = true
-			break
-		}
+	return serveWebhook(gh)
+}
+
+func autoregister(gh *github.Client) {
```

Comment:
Do you want to be able to stop this (stop channel or context)?

Reply:
I'm hoping to add graceful shutdown (using this code: https://github.com/heptio/shutdown/pull/1) in a followup PR.